### ISSUE TITLE
Multiple runs vs one model run leads to different results

### ISSUE THREAD

**Issue by leachader** [huggingface/setfit#381]
Hello Team,

Hope you are doing well.

I was trying to test the effect of adding more samples to the training set (testing for k= [4, 8, 12, 16, 24]) on the performance of Setfit (using F1 as an evaluation metric). In order to do that, I ensure that 1) For every run I have the same training and testing sets by setting a seed of 24 2) That for k=8 samples I have the same 4 samples that are in k=4 and 4 new ones, for k=12 that I have the same 8 samples as in k=8 and 4 new ones etc.. 3) that the testing set is always the same for all values of K to be able to compare the F1 score 4) I set the seed in the trainer and set a model_init to ensure reproducability (see below the model_init function I use)

Case 1: I run a setfit model with K=4, I don't restart the kernel and rerun it for K=8, I get an F1 score of 0.79
Case 2: I initialize the kernel and I run a setfit model once with K=8, ensuring that the training and testing sets are exactly the same as in case 1 and I get an F1 score of 0.82

Note: I am able to reproduce both cases every time I restart the kernel.

I am not sure how to explain this discrepancy, any thoughts?

Looking forward to hearing back from you,

Thank you,
Lea

def model_init():
    return SetFitModel.from_pretrained(
        model_id,
        use_differentiable_head=True,
        head_params={"out_features": num_classes})

**Comment by tomaarsen** [comment]
Hello!

This is quite an interesting experiment that you've prepared. I think it's quite possible that in case 1, the model learned a suboptimal model because of the K=4, and then didn't get as strong when adding the K=8 as just going straight from the pretrained Sentence Transformer with the K=8 (i.e. case 2). 

As an analogy, consider the original Sentence Transformer (i.e. the model_id) to be a pristine piece of paper. You then ask a 4-year old to fold an origami swan with it. Upon finishing it, you tell an 8-year old to also fold a swan with it with that same piece of paper. It'll be a better swan than the 4-year old had made, but you can still tell that the paper was affected by the 4-year old.

If the 8-year old is now given a pristine piece of paper, they would have probably made a slightly better swan than before.

I hope that sort of makes sense.

That all said, I wouldn't have been able to tell you beforehand which of the two cases would be superior, and I think it's very interesting that you've found out that case 2 is superior.

- Tom Aarsen
